From adacc64e006e5ed9411c614a312debdabb352382 Mon Sep 17 00:00:00 2001
From: Yehuda Yitschak <yehuday@marvell.com>
Date: Thu, 4 Feb 2016 10:55:50 +0200
Subject: [PATCH 0060/1345] ahci: add support for CP-110 port address WA

commit  0e80670190b98bda7c510b5db4e81c071e26d47d from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

in CP-110 AHCI unit the port register offsets are not
according to AHCI specification. This patch enables
specifying the port base and offset using device tree
properties

Change-Id: Iada6a27259c03972d5e9e8f03a5356a4aa419ebd
Signed-off-by: Yehuda Yitschak <yehuday@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/27222
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Shadi Ammouri <shadi@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/ata/Kconfig      |   10 ++++++++++
 drivers/ata/ahci.h       |    8 ++++++++
 drivers/ata/ahci_mvebu.c |    5 +++++
 3 files changed, 23 insertions(+)

diff --git a/drivers/ata/Kconfig b/drivers/ata/Kconfig
index de3eaf0..10da93e 100644
--- a/drivers/ata/Kconfig
+++ b/drivers/ata/Kconfig
@@ -171,6 +171,16 @@ config AHCI_OCTEON
 
 	  If unsure, say N.
 
+config CP110_SATA_ADDR_WA
+	bool "Enable WA for CP110 port addresses"
+	depends on AHCI_MVEBU
+	default n
+	help
+	  This option adds a WA for the bad
+	  addresses of CP-110 port registers.
+	  Use this option when using the
+	  CP-110-A0 AHCI ports.
+
 config AHCI_SUNXI
 	tristate "Allwinner sunxi AHCI SATA support"
 	depends on ARCH_SUNXI
diff --git a/drivers/ata/ahci.h b/drivers/ata/ahci.h
index 5db6ab2..71969f2 100644
--- a/drivers/ata/ahci.h
+++ b/drivers/ata/ahci.h
@@ -358,6 +358,10 @@ struct ahci_host_priv {
 	 * be overridden anytime before the host is activated.
 	 */
 	void			(*start_engine)(struct ata_port *ap);
+#ifdef CONFIG_CP110_SATA_ADDR_WA
+	u32			port_base;	/* Offset of ports registers */
+	u32			port_offset;	/* Offset between port registers */
+#endif
 	irqreturn_t 		(*irq_handler)(int irq, void *dev_instance);
 
 	/* only required for per-port MSI(-X) support */
@@ -422,6 +426,10 @@ static inline void __iomem *__ahci_port_base(struct ata_host *host,
 	struct ahci_host_priv *hpriv = host->private_data;
 	void __iomem *mmio = hpriv->mmio;
 
+#ifdef CONFIG_CP110_SATA_ADDR_WA
+	if (hpriv->port_base && hpriv->port_offset)
+		return mmio + hpriv->port_base + (port_no * hpriv->port_offset);
+#endif
 	return mmio + 0x100 + (port_no * 0x80);
 }
 
diff --git a/drivers/ata/ahci_mvebu.c b/drivers/ata/ahci_mvebu.c
index de7128d..cc0cd58 100644
--- a/drivers/ata/ahci_mvebu.c
+++ b/drivers/ata/ahci_mvebu.c
@@ -122,6 +122,11 @@ static int ahci_mvebu_probe(struct platform_device *pdev)
 		ahci_mvebu_regret_option(hpriv);
 	}
 
+#ifdef CONFIG_CP110_SATA_ADDR_WA
+	of_property_read_u32(pdev->dev.of_node, "port_base", &hpriv->port_base);
+	of_property_read_u32(pdev->dev.of_node, "port_offset", &hpriv->port_offset);
+#endif
+
 	rc = ahci_platform_init_host(pdev, hpriv, &ahci_mvebu_port_info,
 				     &ahci_platform_sht);
 	if (rc)
-- 
1.7.9.5

